### PR TITLE
ProductUpgrader: only write newest version available

### PR DESCRIPTION
**Pull request by jamill**
When an upgrade is available, GVFS will download the update in the
background, and git commands will periodically check for the existence
of the downloaded package in a well known directory to advertise that an
update is available.

However, this downloaded update is not used for the actual update.

This changes the logic to write out a file in a well known location
containing the version of the latest available update, and avoid
downloading the actual update.

While this change stands on its own, it will be even more applicable
with the NuGet Upgrader, which will still use GitHub releases to check
for available updates, but will download the actual update from another
location.

### CODE REVIEW

File: GVFS/GVFS.Hooks/Program.cs [hunk 1]
```diff
@@ -112,7 +112,7 @@ private static void RemindUpgradeAvailable()
             int reminderFrequency = 10;
             int randomValue = random.Next(0, 100);
 
-            if (randomValue <= reminderFrequency && ProductUpgraderInfo.IsLocalUpgradeAvailable(GVFSHooksPlatform.GetInstallerExtension()))
+            if (randomValue <= reminderFrequency && ProductUpgraderInfo.IsLocalUpgradeAvailable(tracer: null))
```
Comment (mjcheetham) [inline GVFS/GVFS.Hooks/Program.cs]: The hooks don't have a tracer instance?

File: GVFS/GVFS.Service/ProductUpgradeTimer.cs [hunk 2]
```diff
@@ -23,7 +22,7 @@ public void Start()
         {
             if (!GVFSEnlistment.IsUnattended(this.tracer))
             {
-                TimeSpan startTime = TimeSpan.Zero;
+                TimeSpan startTime = TimeSpan.FromMinutes(1);
```
Comment (mjcheetham) [inline GVFS/GVFS.Service/ProductUpgradeTimer.cs]: What is the reason for this change?

File: GVFS/GVFS.Service/ProductUpgradeTimer.cs [hunk 3]
```diff
@@ -53,70 +52,101 @@ public void Dispose()
...
             {
-                Version newerVersion = null;
-                string detailedError = null;
-                if (!productUpgrader.UpgradeAllowed(out errorMessage))
+                try
                 {
-                    return false;
-                }
+                    // The upgrade check always goes against GitHub
+                    GitHubUpgrader productUpgrader = GitHubUpgrader.Create(
```
Comment (mjcheetham) [inline GVFS/GVFS.Service/ProductUpgradeTimer.cs]: Do we expect the `GitHubUpgrader` instance to be the same each time the timer fires? Could we not keep this instance around for the lifetime of the `ProductUpgraderTimer` rather than each `TimerCallback(..)`?

File: GVFS/GVFS.Service/ProductUpgradeTimer.cs [hunk 4]
```diff
@@ -53,70 +52,101 @@ public void Dispose()
...
-                    // Make sure there a no asset installers remaining in the Downloads directory. This can happen if user
-                    // upgraded by manually downloading and running asset installers.
-                    ProductUpgraderInfo.DeleteAllInstallerDownloads();
-                    errorMessage = null;
-                    return true;
-                }
+                        return;
+                    }
 
-                if (productUpgrader.TryDownloadNewestVersion(out detailedError))
```
Comment (mjcheetham) [inline GVFS/GVFS.Service/ProductUpgradeTimer.cs]: So when do we actually download the packages now? Also what would happen if the version number in the 'latest' file is different from the actual latest version on the server/feed? Do we download the version number as specified in the file or just 'latest'?

(Just thinking what would happen in the case we pulled an update file but a user already did the check and has a 'latest' file pointing to that now revoked version.)

File: GVFS/GVFS.Hooks/Program.cs (hunk 1 above)
Comment (jamill) [reply GVFS/GVFS.Hooks/Program.cs]: I will check again - but I did not see any tracers from the hooks. I don't recall us having any tracing from the hooks themselves for individual git commands...

File: GVFS/GVFS.Service/ProductUpgradeTimer.cs (hunk 2 above)
Comment (jamill) [reply GVFS/GVFS.Service/ProductUpgradeTimer.cs]: Originally - I added a 1 minute delay before GVFS service checked for updates so that I had time to attach the debugger to the service. I tried following the basic tricks to debug service startup, but I was not successful - I suspect I was not getting interactive prompts, even though I had set the bit to allow the service to interact with the desktop....

However, I thought this might not be a bad change in general, as we delaying the initial check by a minute has minimal (no?) downside, and allows us to attach a debugger if needed. It could be a positive to wait a minute so we don't check while starting up.

I can put this back, if people object.

File: GVFS/GVFS.Service/ProductUpgradeTimer.cs (hunk 4 above)
Comment (jamill) [reply GVFS/GVFS.Service/ProductUpgradeTimer.cs]: > So when do we actually download the packages now?

We download the package when running the Upgrade command.

> Also what would happen if the version number in the 'latest' file is different from the actual latest version on the server/feed? Do we download the version number as specified in the file or just 'latest'?

We will always download the latest. If the `latest` version changes in between the version in the file being updated and running the Upgrade command, we will download and install the new latest. This has been brought up as something we might want to tweak, so that someone can download and install a specific version to avoid race conditions such as this, but we don't have that yet.

File: GVFS/GVFS.Service/ProductUpgradeTimer.cs (hunk 3 above)
Comment (jamill) [reply GVFS/GVFS.Service/ProductUpgradeTimer.cs]: I suppose the configured ring could change after the service has started. If we went this route, changing the GVFS config would require a service restart to pick up the new values...

File: GVFS/GVFS.Service/ProductUpgradeTimer.cs (hunk 4 above)
Comment (mjcheetham) [reply GVFS/GVFS.Service/ProductUpgradeTimer.cs]: > We download the package when running the Upgrade command.

But from the changes I see in the PR so far the 'download' code has been removed, not moved. Are you saying the actual call to 'download' is already in the Upgrade command code, and previously we were downloading twice (once for the check, once for the actual upgrade cmd)?

> This has been brought up as something we might want to tweak, so that someone can download and install a specific version to avoid race conditions such as this, but we don't have that yet.

..or at least show a warning/message that "the latest version is now n+1".

File: GVFS/GVFS.Service/ProductUpgradeTimer.cs (hunk 3 above)
Comment (mjcheetham) [reply GVFS/GVFS.Service/ProductUpgradeTimer.cs]: > If we went this route, changing the GVFS config would require a service restart to pick up the new values...

Right, I suspected this might have been the reason for not doing it. Thanks for confirming! 😃

File: GVFS/GVFS.Common/ProductUpgraderInfo.cs [hunk 5]
```diff
@@ -10,39 +10,59 @@ public class ProductUpgraderInfo
...
+                    return false;
+                }
 
-        public static bool IsLocalUpgradeAvailable(string installerExtension)
-        {
-            string downloadDirectory = GetAssetDownloadsPath();
-            if (Directory.Exists(downloadDirectory))
+                Version currentVersion = new Version(CurrentGVFSVersion());
+                return highestAvailableVersion > currentVersion;
+            }
```
Comment (alameenshah) [inline GVFS/GVFS.Common/ProductUpgraderInfo.cs]: This looks a lot cleaner than searching for files matching version numbers.

File: GVFS/GVFS.Common/ProductUpgraderInfo.cs [hunk 6]
```diff
@@ -104,5 +129,37 @@ private static void RecursiveDelete(string path)
...
+        {
+            Version highestAvailableVersion = null;
+            string highestAvailableVersionFile = GetHighestAvailableVersionFilePath();
+
+            if (!File.Exists(highestAvailableVersionFile))
+            {
+                return null;
+            }
+
+            string contents = File.ReadAllText(highestAvailableVersionFile).Trim();
```
Comment (alameenshah) [inline GVFS/GVFS.Common/ProductUpgraderInfo.cs]: Nit - Do we need to read the whole file? Should we stop at the first non-empty line. (Not a big issue since the highestAvailableVersionFile is not world writable.)

**Comment by alameenshah** [comment]
Functional tests for Upgrade reminder messaging has to be updated to use the new `highestAvailableVersion` file.

File: GVFS/GVFS.Hooks/Program.cs (hunk 1 above)
Comment (wilbaker) [reply GVFS/GVFS.Hooks/Program.cs]: That's correct, the hooks don't have any tracing in them.

File: GVFS/GVFS.Common/ProductUpgraderInfo.cs [hunk 7]
```diff
@@ -10,39 +10,59 @@ public class ProductUpgraderInfo
...
+                        CreateEventMetadata(ex),
+                        "Exception encountered when determining if an upgrade is available.");
                 }
             }
 
             return false;
         }
 
-        public static IEnumerable<string> PossibleGVFSInstallerNamePrefixes()
+        public static void RecordHighestAvailableVersion(Version highestAvailableVersion)
```
Comment (wilbaker) [inline GVFS/GVFS.Common/ProductUpgraderInfo.cs]: Use `PhysicalFileSystem` rather than `System.IO` so that we can unit test this method

File: GVFS/GVFS.Common/ProductUpgraderInfo.cs [hunk 8]
```diff
@@ -10,39 +10,59 @@ public class ProductUpgraderInfo
...
+                {
+                    return false;
+                }
 
-        public static bool IsLocalUpgradeAvailable(string installerExtension)
-        {
-            string downloadDirectory = GetAssetDownloadsPath();
-            if (Directory.Exists(downloadDirectory))
+                Version currentVersion = new Version(CurrentGVFSVersion());
+                return highestAvailableVersion > currentVersion;
```
Comment (wilbaker) [inline GVFS/GVFS.Common/ProductUpgraderInfo.cs]: We previously weren't comparing the downloaded versions against our own process version.  Would it be enough to just test for the existence of a marker file here?

File: GVFS/GVFS.Common/ProductUpgraderInfo.cs [hunk 9]
```diff
@@ -104,5 +129,37 @@ private static void RecursiveDelete(string path)
 
             directory.Delete();
         }
+
+        private static Version ReadHighestRecordedAvailableVersion()
```
Comment (wilbaker) [inline GVFS/GVFS.Common/ProductUpgraderInfo.cs]: Similar comment here, use `PhysicalFileSystem` and add unit tests.

File: GVFS/GVFS.Service/ProductUpgradeTimer.cs [hunk 10]
```diff
@@ -53,70 +52,101 @@ public void Dispose()
...
-                if (!productUpgrader.UpgradeAllowed(out errorMessage))
+                try
                 {
-                    return false;
-                }
+                    // The upgrade check always goes against GitHub
+                    GitHubUpgrader productUpgrader = GitHubUpgrader.Create(
+                        this.tracer,
+                        false,
+                        false,
```
Comment (wilbaker) [inline GVFS/GVFS.Service/ProductUpgradeTimer.cs]: Minor: Add named parameters here so that it's easier for readers to see what parameters are being set to `false`

File: GVFS/GVFS.Service/ProductUpgradeTimer.cs [hunk 11]
```diff
@@ -53,70 +52,101 @@ public void Dispose()
...
-                if (newerVersion == null)
-                {
-                    // Already up-to-date
-                    // Make sure there a no asset installers remaining in the Downloads directory. This can happen if user
-                    // upgraded by manually downloading and running asset installers.
-                    ProductUpgraderInfo.DeleteAllInstallerDownloads();
-                    errorMessage = null;
-                    return true;
-                }
+                        return;
```
Comment (wilbaker) [inline GVFS/GVFS.Service/ProductUpgradeTimer.cs]: Should we also log something if `errorMessage` is `null`? How would we get in that situation?

File: GVFS/GVFS.Service/ProductUpgradeTimer.cs [hunk 12]
```diff
@@ -53,70 +52,101 @@ public void Dispose()
...
+                    }
 
-                if (productUpgrader.TryDownloadNewestVersion(out detailedError))
-                {
-                    errorMessage = null;
-                    return true;
+                    InstallerPreRunChecker prerunChecker = new InstallerPreRunChecker(this.tracer, string.Empty);
+                    if (!prerunChecker.TryRunPreUpgradeChecks(out _))
+                    {
+                        return;
```
Comment (wilbaker) [inline GVFS/GVFS.Service/ProductUpgradeTimer.cs]: Should we trace the output parameter?

File: GVFS/GVFS.Service/ProductUpgradeTimer.cs [hunk 13]
```diff
@@ -53,70 +52,101 @@ public void Dispose()
...
+                    }
+
+                    if (!productUpgrader.UpgradeAllowed(out errorMessage))
+                    {
+                        if (errorMessage != null)
+                        {
+                            this.tracer.RelatedError(errorMessage);
```
Comment (wilbaker) [inline GVFS/GVFS.Service/ProductUpgradeTimer.cs]: If `UpgradeAllowed` returns `false`, is it always an error?  I am wondering if any of these should actually be `RelatedWarning`.

File: GVFS/GVFS.Service/ProductUpgradeTimer.cs [hunk 14]
```diff
@@ -53,70 +52,101 @@ public void Dispose()
...
+                        return;
+                    }
+
+                    ProductUpgraderInfo.RecordHighestAvailableVersion(newerVersion);
                 }
-                else
+                catch (Exception ex) when (
+                    ex is IOException ||
+                    ex is UnauthorizedAccessException ||
+                    ex is NotSupportedException)
```
Comment (wilbaker) [inline GVFS/GVFS.Service/ProductUpgradeTimer.cs]: Should we be catching all exception types here?  

It looks like if some other exception was thrown the background thread would die but the service would keep running.

File: GVFS/GVFS.Service/ProductUpgradeTimer.cs [hunk 15]
```diff
@@ -53,70 +52,101 @@ public void Dispose()
...
+            errorMessage = null;
+            tracer.RelatedInfo("Querying server for latest version...");
+
+            if (!productUpgrader.TryQueryNewestVersion(out newVersion, out string detailedError))
+            {
+                errorMessage = "Could not fetch new version info. " + detailedError;
+                return false;
+            }
+
+            string logMessage = newVersion is null ? "No newer versions available." : $"Newer version available: {newVersion}.";
```
Comment (wilbaker) [inline GVFS/GVFS.Service/ProductUpgradeTimer.cs]: Minor: I think this came up in another PR as well, but elsewhere in the codebase we're using ` == null` to check if a variable is `null` and we should stay consistent.

File: GVFS/GVFS.Service/ProductUpgradeTimer.cs [hunk 16]
```diff
@@ -53,70 +52,101 @@ public void Dispose()
...
+                            productUpgrader,
+                            out Version newerVersion,
+                            out errorMessage))
+                    {
+                        if (errorMessage != null)
+                        {
+                            this.tracer.RelatedError(errorMessage);
+                        }
+
+                        return;
```
Comment (wilbaker) [inline GVFS/GVFS.Service/ProductUpgradeTimer.cs]: In this case (we're on the latest version) should the `HighestAvailableVersion` file be deleted?

File: GVFS/GVFS.Service/ProductUpgradeTimer.cs (hunk 2 above)
Comment (wilbaker) [reply GVFS/GVFS.Service/ProductUpgradeTimer.cs]: I don't see a problem with waiting a minute.

**Comment by wilbaker** [comment]
@jamill I just noticed this PR is into master, should it be into the feature branch so that we can test all of the upgrader changes together?

File: GVFS/GVFS.Common/ProductUpgraderInfo.cs (hunk 8 above)
Comment (jamill) [reply GVFS/GVFS.Common/ProductUpgraderInfo.cs]: Probably - the benefit of reading the version here is we can confirm that we are out-of-date.

The window of when this file would exist, but the user is actually up-to-date is pretty small - in particular it could happen when a user manually installs a new version (not via upgrade), but the background process has not cleared the file yet.

That brings up a point I missed - upgrade should clear this file when it runs... Although as is, the user would not see incorrect results due to this check.

**Comment by jamill** [comment]
> @jamill I just noticed this PR is into master

That was intentional - but I should have called it out. As this change stands on its own, I targeted master. I don't mind targeting the feature branch, but my thought was this keeps the feature branch smaller... On the flip side, it could introduce churn in the feature branch as it will need to absorb these changes, but I felt this was the easier path overall.

I can target either branch if people prefer otherwise.

File: GVFS/GVFS.Common/ProductUpgraderInfo.cs (hunk 9 above)
Comment (mjcheetham) [reply GVFS/GVFS.Common/ProductUpgraderInfo.cs]: Would bringing in and initialising these dependencies increase the spin-up time for the hooks?

File: GVFS/GVFS.Service/ProductUpgradeTimer.cs (hunk 13 above)
Comment (jamill) [reply GVFS/GVFS.Service/ProductUpgradeTimer.cs]: I made most of these RelatedWarning

File: GVFS/GVFS.Common/ProductUpgraderInfo.cs (hunk 6 above)
Comment (jamill) [reply GVFS/GVFS.Common/ProductUpgraderInfo.cs]: Good point - as the file is controlled by us, I didn't stop at the 1st line. If someone is tampering with the file, then we have bigger problems... @wilbaker suggested to not even read the contents of the file, which could also work...

File: GVFS/GVFS.Service/ProductUpgradeTimer.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jamill) [reply GVFS/GVFS.Service/ProductUpgradeTimer.cs]: I could - something like `TryCreateUpgraderForAnonymousPolling`?

File: GVFS/GVFS.Common/ProductUpgraderInfo.cs (hunk 6 above)
Comment (wilbaker) [reply GVFS/GVFS.Common/ProductUpgraderInfo.cs]: If we decide to continue to read the contents of the file I think it's okay to read the entire file.  There should only be one line, and if we read in unexpected data it seems better to have `Version.TryParse` fail than silently ignore the unexpected extra data.

File: GVFS/GVFS.Common/ProductUpgraderInfo.cs (hunk 8 above)
Comment (wilbaker) [reply GVFS/GVFS.Common/ProductUpgraderInfo.cs]: > Probably - the benefit of reading the version here is we can confirm that we are out-of-date.

That makes sense.  I wasn't sure how much extra overhead we're adding to the hooks with  open+read+parse vs a file existence check.

File: GVFS/GVFS.Common/ProductUpgraderInfo.cs (hunk 9 above)
Comment (wilbaker) [reply GVFS/GVFS.Common/ProductUpgraderInfo.cs]: Rather than having hooks consume `ProductUpgraderInfo.cs` directly, we could move the method(s) the hooks need to a new `ProductUpgraderInfo.Shared.cs`, and then methods like this one could use `PhysicalFileSystem` without having to pull `PhysicalFileSystem` into the hooks project.

File: GVFS/GVFS.Service/ProductUpgradeTimer.cs [hunk 17]
```diff
@@ -53,70 +54,102 @@ public void Dispose()
...
+            string logMessage = newVersion == null ? "No newer versions available." : $"Newer version available: {newVersion}.";
+            tracer.RelatedInfo(logMessage);
+
+            return true;
+        }
+
+        private bool TraceException(Exception ex)
+        {
+            EventMetadata metadata = this.CreateEventMetadata(ex);
+            this.tracer.RelatedError(metadata, message);
```
[the remark this answers is not in the document]
Comment (wilbaker) [reply GVFS/GVFS.Service/ProductUpgradeTimer.cs]: It looks like this comment crossed with your `fixup` commit

[the remark this answers is not in the document]
Comment (jamill) [reply GVFS/GVFS.Service/ProductUpgradeTimer.cs]: Yeah - unfortunately this code is not compiled on non windows platforms at the moment.

File: GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GVFSUpgradeReminderTests.cs [hunk 18]
```diff
@@ -77,15 +76,12 @@ private void EmptyDownloadDirectory()
...
-            string gitInstallerPath = Path.Combine(this.upgradeDirectory, GitInstallerName);
+            string gvfsUpgradeAvailableFilePath = Path.Combine(this.upgradeDirectory, HighestAvailableVersionFileName);
 
             this.EmptyDownloadDirectory();
 
-            this.fileSystem.CreateEmptyFile(gvfsInstallerPath);
-            this.fileSystem.CreateEmptyFile(gitInstallerPath);
-            this.fileSystem.FileExists(gvfsInstallerPath).ShouldBeTrue();
-            this.fileSystem.FileExists(gitInstallerPath).ShouldBeTrue();
+            this.fileSystem.CreateEmptyFile(gvfsUpgradeAvailableFilePath);
```
[the remark this answers is not in the document]
Comment (jamill) [reply GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GVFSUpgradeReminderTests.cs]: We do not have any code that reads in the contents of this file, only that it exists, so I was treating this as an implementation detail. If / when it becomes important, we could add a test then. We could alternatively just write a blank file...

[the remark this answers is not in the document]
Comment (wilbaker) [reply GVFS/GVFS.FunctionalTests/Tests/EnlistmentPerFixture/GVFSUpgradeReminderTests.cs]: > We could alternatively just write a blank file...

I think there's value in writing out the version from a support perspective.

> We do not have any code that reads in the contents of this file, only that it exists

That's a good point, we can always revisit this if we decide to start reading the value.

File: GVFS/GVFS.Service/ProductUpgradeTimer.cs [hunk 19]
```diff
@@ -53,70 +54,102 @@ public void Dispose()
...
+                    ex is NotSupportedException)
                 {
-                    errorMessage = null;
-                    return true;
+                    this.tracer.RelatedWarning(
+                        CreateEventMetadata(ex),
+                        "Exception encountered recording highest available version");
                 }
-                else
+                catch (Exception ex) when (this.TraceException(ex))
```
[the remark this answers is not in the document]
Comment (jamill) [reply GVFS/GVFS.Service/ProductUpgradeTimer.cs]: > Is there any reason we can't do that here?

We can do that here - I think the approach has several drawbacks which can make it harder to debug, but we can follow up another time. I will update the code to be consistent.

File: GVFS/GVFS.UnitTests/Common/ProductUpgraderInfoTests.cs [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+            this.mockFileSystem = null;
+            this.productUpgraderInfo = null;
+            this.tracer = null;
+        }
+
+        [TestCase]
+        public void RecordHighestVersion()
+        {
+            this.productUpgraderInfo.RecordHighestAvailableVersion(new Version("1.0.0.0"));
+            this.mockFileSystem.Verify(fileSystem => fileSystem.WriteAllText(this.expectedNewVersionExistsFilePath, It.IsAny<string>()), Times.Once());
```
[the remark this answers is not in the document]
Comment (jamill) [reply GVFS/GVFS.UnitTests/Common/ProductUpgraderInfoTests.cs]: I think this is a good idea - however, as we do not read the version from this file, I will put this on my list of items to follow up on, rather than include it in this PR.

File: GVFS/GVFS.Service/ProductUpgradeTimer.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (jamill) [reply GVFS/GVFS.Service/ProductUpgradeTimer.cs]: I like this suggestion - but was going to hold off on implementing it in this PR. We can come back after we have the different streams of work merged, to reduce churn / conflicts.

File: GVFS/GVFS.UnitTests/Common/ProductUpgraderInfoTests.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (wilbaker) [reply GVFS/GVFS.UnitTests/Common/ProductUpgraderInfoTests.cs]: > I will put this on my list of items to follow up on, rather than include it in this PR.

Could you create a GitHub issue that has planned follow up work items?

[the remark this answers is not in the document]
Comment (wilbaker) [reply GVFS/GVFS.UnitTests/Common/ProductUpgraderInfoTests.cs]: You could also consider renaming the test for now to something like `CreatesHighestVersionFile` as the test is not validating that the highest version was recorded.
